app/src/main/java/com/ikarabulut/RulesFactory.java

```diff
@@ -6,6 +6,8 @@ public RuleSet getRules(int ruleSelection) throws InvalidRuleSelectionException
         return switch (ruleSelection) {
             case 1 -> new DefaultRules();
             case 2 -> new ComeAliveWithFour();
+            case 3 -> new ComeAliveWithTwo();
```

Comment:
Note: one benefit of the Factory pattern and polymorphism is that we had to only modify one place to allow the user to select more options.  That logic isn't smeared across many places in the program.

Reply:
I spent about an hr trying to refactor this following Ch 7 of 99 bottles which disperses the choosing logic into an Array (in ruby, which it is a lot easier to do this in Ruby than in Java) by turning the superclass `RuleSet` into the _factory_. 

I was able to get to creating a selector method which made a `List.of({each subclass})' and then doing a `for` loop through that `List` and then selecting the class that used that number in a `handles()` method.

I could have seen myself spending another hour minimum on it to try and get something working and didn't want to pull too much of my time away from the echo server.

Maybe we can pair this Friday on this problem to make it work?

Reply:
Oh, interesting.  

Reviewing Chapter 7, this passage is a good summary of the message I've been trying to make about polymorphism and encapsulation:

> “From the message sender’s point of view, all players of a role are exactly the same. Message senders depend on the role player’s exposed intentions while remaining studiously ignorant of their detailed internal implementations. They know what their collaborators do, but refuse to be aware of how they do it.

> A system comprised of message senders who collaborate with role-playing objects can be extremely tolerant of unexpected change. For example, imagine that you create a new class that represents a new variant of bottle number. This new bottle number is interchangeable with every other one, so any class that already collaborates with an existing bottle number will be able to seamlessly collaborate with this new one. The message sender doesn’t have to change; from its point of view this new bottle number is the same as all the others.”

That applies to the Rules interface we've created in Game of Life.

Reply:
> I spent about an hr trying to refactor this following Ch 7 of 99 bottles which disperses the choosing logic into an Array (in ruby, which it is a lot easier to do this in Ruby than in Java) by turning the superclass RuleSet into the factory.

In this case, RulesFactory can be considered technically "closed" because if we add a new RuleSet class, then we have to open up RulesFactory to teach it about the new class.

In this case, that trade-off seems worth it to me.  The programmer must specify a RuleSet within the program before it can be used. It cannot be supplied by the user, for example.  When a new RuleSet is added, we only have to update one place in the code to allow it to be used.

The trade-offs change based on how often new RuleSets get added.  If RuleSets are added frequently, then a more dynamic approach will save time and be worth it. Then we adapt the techniques we use based on the places the code gets changed the most.

> I was able to get to creating a selector method which made a List.of({each subclass})' and then doing a forloop through that `List` and then selecting the class that used that number in `handles()` method.

The downside of choosing the registration technique in this context is that what should `handles` method check for?  If you make it the integer the user selects, then a RuleSet knows about something that Game or RulesFactory should be managing.  If I add a new RuleSet, how do I know which integers are free to use?  What if the order needs to be changed?

Reply:
For my favorite example of the Open/Closed principle, check out how Rails middlewares work and can be configured.

https://guides.rubyonrails.org/rails_on_rack.html

A middleware modifies HTTP incoming requests and outgoing responses. The classes can be supplied by the user, and they can do arbitrary things with the HTTP request and response. So the Rails developers don't know ahead of time what a user will want to do.

Here's the simplest example of a Middleware.  This one adds an HTTP header, `X-Request-Id`, and assigns it a random value:
https://github.com/rails/rails/blob/3872bc0e54d32e8bf3a6299b0bfe173d94b072fc/actionpack/lib/action_dispatch/middleware/request_id.rb#L17

Specifically check out the section in the documentation on how Middlewares are added and removed from Rails:
https://guides.rubyonrails.org/rails_on_rack.html#configuring-middleware-stack

Rails uses a registry pattern for their factory. Why?  Because the Rails developers got enough requests to change it, so that it made sense to give users control over how the list of middlewares are configured.

* `config.middleware.use(new_middleware, args)` - Adds the new middleware at the bottom of the middleware stack.
* `config.middleware.insert_before(existing_middleware, new_middleware, args)` - Adds the new middleware before the specified existing middleware in the middleware stack.
* `config.middleware.insert_after(existing_middleware, new_middleware, args)` - Adds the new middleware after the specified existing middleware in the middleware stack.

Rails users can add, remove, and reorder middlewares in their applications and the Rails developers don't have to know about it. You don't have to file a GitHub Issue in Rails to get your middleware added to their list before you can use it in your application.

The Middleware Pattern (my name for it) is a canonical example of Open/Closed. It is open for extension, but closed for modification. You can add new middlewares, and you don't edit any source code in Rails to tell it about the new middleware that you want to add (or remove).  This pattern is common to most web frameworks, not just Rails. But, since you know Rails, I figure this example will be the easier to follow.